impeller/entity/entity_pass.cc

```diff
@@ -174,146 +178,212 @@ bool EntityPass::Render(ContentContext& renderer,
-    // Setup entity element for rendering ======================================
-    // =========================================================================
-    if (const auto& entity = std::get_if<Entity>(&element)) {
-      element_entity = *entity;
-      if (!position.IsZero()) {
-        // If the pass image is going to be rendered with a non-zero position,
-        // apply the negative translation to entity copies before rendering them
-        // so that they'll end up rendering to the correct on-screen position.
-        element_entity.SetTransformation(
-            Matrix::MakeTranslation(Vector3(-position)) *
-            element_entity.GetTransformation());
-      }
+    if (subpass->delegate_->CanElide()) {
+      return EntityPass::EntityResult::Skip();
     }
 
-    // =========================================================================
-    // Setup subpass element for rendering =====================================
-    // =========================================================================
-    else if (const auto& subpass_ptr =
-                 std::get_if<std::unique_ptr<EntityPass>>(&element)) {
-      auto subpass = subpass_ptr->get();
-
-      if (subpass->delegate_->CanElide()) {
-        continue;
+    if (subpass->delegate_->CanCollapseIntoParentPass() &&
+        !subpass->backdrop_filter_proc_.has_value()) {
+      // Directly render into the parent target and move on.
+      if (!subpass->OnRender(renderer, pass_context.GetRenderTarget(), position,
+                             position, stencil_depth_floor)) {
+        return EntityPass::EntityResult::Failure();
       }
+      return EntityPass::EntityResult::Skip();
+    }
 
-      if (subpass->delegate_->CanCollapseIntoParentPass()) {
-        // Directly render into the parent target and move on.
-        if (!subpass->RenderInternal(renderer, render_target, position,
-                                     pass_depth, stencil_depth_floor)) {
-          return false;
-        }
-        continue;
-      }
+    std::optional<Rect> subpass_coverage;
 
-      const auto subpass_coverage = GetSubpassCoverage(*subpass);
+    if (subpass->elements_.empty() && backdrop_filter_proc_.has_value()) {
+      // An empty subpass with a backdrop filter should render the backdrop over
+      // the entire coverage of the current pass.
+      subpass_coverage = Rect(
+          position, Size(pass_context.GetRenderTarget().GetRenderTargetSize()));
+    } else {
+      subpass_coverage = GetSubpassCoverage(*subpass);
       if (!subpass_coverage.has_value()) {
-        continue;
+        return EntityPass::EntityResult::Skip();
       }
 
       if (subpass_coverage->size.IsEmpty()) {
         // It is not an error to have an empty subpass. But subpasses that can't
         // create their intermediates must trip errors.
-        continue;
+        return EntityPass::EntityResult::Skip();
```

Comment:
Is it possible for the subpass to be non-"empty", but produce an empty coverage? If so, then we don't want to skip here if there is a backdrop filter.

Reply:
Another way to look at this is that the subpass has some coverage, and if there is an optional backdrop filter, then it also has some coverage, and then you combine the two (and clip the combined coverage). Now, if that combined clipped coverage is null, then it's OK to skip.

Reply:
Fixed.